### ISSUE TITLE
Dynamic routes: trash in the url

### ISSUE THREAD

**Issue by yaleksandr89** [prestaconcept/PrestaSitemapBundle#238]
Hi there!

My setting:
`services.yaml`
```yaml
  app.sitemap_subscriber:
    class: App\Application\Cms\EventListener\SitemapListener
    arguments:
      - "@router"
      - '@App\Application\Sonata\NewsBundle\Entity\PostRepository'
    tags:
      - { name: "kernel.event_subscriber", priority: 100 }
```
`SitemapListener`
```php
<?php

namespace App\Application\Cms\EventListener;

use App\Application\Sonata\NewsBundle\Entity\PostRepository;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
use Presta\SitemapBundle\Event\SitemapPopulateEvent;
use Presta\SitemapBundle\Service\UrlContainerInterface;
use Presta\SitemapBundle\Sitemap\Url\UrlConcrete;

class SitemapListener implements EventSubscriberInterface
{
    /**
     * @var UrlGeneratorInterface
     */
    private $urlGenerator;

    /**
     * @var PostRepository
     */
    private $postRepository;

    /**
     * @param UrlGeneratorInterface $urlGenerator
     * @param PostRepository $postRepository
     */
    public function __construct(UrlGeneratorInterface $urlGenerator, PostRepository $postRepository)
    {
        $this->urlGenerator = $urlGenerator;
        $this->postRepository = $postRepository;
    }

    /**
     * @inheritdoc
     */
    public static function getSubscribedEvents()
    {
        return [
            SitemapPopulateEvent::ON_SITEMAP_POPULATE => 'populate',
        ];
    }

    /**
     * @param SitemapPopulateEvent $event
     */
    public function populate(SitemapPopulateEvent $event): void
    {
        $this->registerBlogPostsUrls($event->getUrlContainer());
    }

    /**
     * @param UrlContainerInterface $urls
     */
    public function registerBlogPostsUrls(UrlContainerInterface $urls): void
    {
        $posts = $this->postRepository->findAll();

        foreach ($posts as $post) {
            $urls->addUrl(
                new UrlConcrete(
                    $this->urlGenerator->generate(
                        'sonata_news_home',
                        ['slug' => $post->getSlug()],
                        UrlGeneratorInterface::ABSOLUTE_URL
                    )
                ),
                'articles'
            );
        }
    }
}
```
When creating sitemap.xml, I get a url (example):
```
<url>
<loc>http://symfony.loc/articles/?slug=article-1</loc>
<lastmod>2020-06-04T21:01:50+03:00</lastmod>
<changefreq>daily</changefreq>
<priority>0.5</priority>
</url>
...
<url>
<loc>http://symfony.loc/articles/?slug=article-100</loc>
<lastmod>2020-06-04T21:01:50+03:00</lastmod>
<changefreq>daily</changefreq>
<priority>0.5</priority>
</url>
```

How to remove from the url: `?slug=` ??? 
I guess that comes from behind:
```php
$urls->addUrl(
                new UrlConcrete(
                    $this->urlGenerator->generate(
                        'sonata_news_home',
                      //HERE--->  ['slug' => $post->getSlug()],
                        UrlGeneratorInterface::ABSOLUTE_URL
                    )
                ),
                'articles'
            );
```

But in the second parameter of the method `generate(){}` need to pass - array. 

How can I solve this problem? Bundle very much liked, there is no special desire to look for an alternative.

Thanks!

**Comment by yann-eugone** [comment]
Hi there.

I'm not sure of what you are trying to achieve, but :

Your problem is definitely the `['slug' => $post->getSlug()]`.
Apparently the route parameter named `'slug'` do not exists on the route named `'sonata_news_home'` => this is why Symfony is adding the `slug=whatever` query parameter.

TL;DR
```diff
- ['slug' => $post->getSlug()]
+ []
```

**Comment by yaleksandr89** [comment]
> Your problem is definitely the ['slug' => $post->getSlug()].

Hi there.

You were right, and my guess was correct. It all came down to my little experience developing on Symfony.

There were two routes:
```bash
  sonata_news_view          /articles/{permalink}.{_format}                                          
  sonata_news_home        /articles/ 
```
I linked to a static route to get the url of all articles: `sonata_news_home`, should i have: `articles/{permalink}.{_format}`. And in this case method `registerBlogPostsUrls(UrlContainerInterface $urls){}`:
```
 /**
     * @param UrlContainerInterface $urls
     */
    public function registerBlogPostsUrls(UrlContainerInterface $urls): void
    {
        $posts = $this->postRepository->findAll();

        foreach ($posts as $post) {
            $urls->addUrl(
                new UrlConcrete(
                    $this->urlGenerator->generate(
                        'sonata_news_view',
                        ['permalink' => $post->getSlug()],
                        UrlGeneratorInterface::ABSOLUTE_URL
                    )
                ),
                'articles'
            );

        }
    }
```

P.S.
To avoid creating a new issue, I will ask here. Please tell me if you need to get information from several dynamic routes, what should I do?

For each route, create: `SitemapListener` (`SitemapListenerArticle`, `SitemapListenerNews` etc)?

**Comment by yann-eugone** [comment]
The `sonata_news_home` is a route without any parameter, it sounds odd tha you are using it in a dynamic url provider.
Using `sonata_news_view` looks like a better idea.

For your other questions, I'm not sure what to answer. If you have several sources for dynamic routes (news, article, blog post, products, ...) you can create 1 listener per source, but this is not required.
As an example, in the sandbox project the listener is registering several urls sources in the same listener : https://github.com/yann-eugone/presta-sitemap-test-project/blob/master/src/Listener/SitemapListener.php

=> This is your call

**Comment by yaleksandr89** [comment]
That's it. Thank you very much for your help!

Good luck!
